// @framer/lingzhou.searcher/code/Searcher.tsx
import * as React from "react"
import { addPropertyControls, ControlType } from "framer"

class Search extends React.Component {
    constructor(props) {
        super(props)
        // enter your search result here, seperated by comma
        const string = "apple,orange,pear"
        const resultsArray = string.split(",")

        this.state = {
            data: resultsArray,
            list: undefined,
        }
    }
    searchData(e) {
        var queryData = []
        if (e.target.value !== "") {
            this.state.data.forEach(function(person) {
                if (person.toLowerCase().indexOf(e.target.value) !== -1) {
                    if (queryData.length < 10) {
                        queryData.push(person)
                    }
                }
            })
        }
        this.setState({ list: queryData })
    }
    render() {
        return (
            <div>
                <SearchBar
                    placeholder={this.props.placeholder}
                    search={this.searchData.bind(this)}
                />
                {this.state.list ? (
                    <SearchResult data={this.state.list} />
                ) : null}
            </div>
        )
    }
}

//L2 components: search bar, search result
class SearchBar extends React.Component {
    render() {
        return (
            <div>
                <input
                    style={{
                        fontSize: "14px",
                        outline: "none",
                        width: "100%",
                        padding: "8px",
                        marginBottom: "8px",
                    }}
                    onChange={this.props.search}
                    placeholder={this.props.placeholder}
                />
            </div>
        )
    }
}

class SearchResult extends React.Component {
    render() {
        return (
            <div>
                <ul
                    style={{
                        backgroundColor: "#fff",
                        borderRadius: "4px",
                        margin: "0px",
                        listStyle: "none",
                        padding: "0px",
                    }}
                >
                    {this.props.data.map(function(value) {
                        return <Item key={value} val={value} />
                    })}
                </ul>
            </div>
        )
    }
}

//L3 compoents: each item of the search result
class Item extends React.Component {
    render() {
        return (
            <li
                style={{
                    padding: "8px",
                }}
            >
                {this.props.val}
            </li>
        )
    }
}
export class Searcher extends React.Component {
    render() {
        return (
            <div style={{}}>
                <Search placeholder={this.props.placeholder} />
            </div>
        )
    }
}

// default values for the module
Searcher.defaultProps = {
    placeholder: "Search...",
}

// use Framer's control panel to modify attributes
addPropertyControls(Searcher, {
    placeholder: {
        title: "Placeholder",
        type: ControlType.String,
        defaultValue: "Search...",
    },
})
