// @framer/framer.default/code/Cursors.tsx
import * as React from "react"
import { addPropertyControls, ControlType } from "framer"
import { MacOSCursors } from "./assets/cursors"
import { defaultEvents } from "./lib/constants"

export function Cursors(props) {
    const { platform, macOSCursor, onClick, onMouseUp, onMouseDown, onMouseEnter, onMouseLeave, width, height } = props
    const Cursor = React.useMemo(() => {
        switch (platform) {
            case "MacOS":
                return MacOSCursors[macOSCursor]
            default:
                console.error(`Unsupported platform: ${platform}`)
                break
        }
    }, [platform, macOSCursor])

    return (
        <div
            style={{ overflow: "hidden", width: "100%", height: "100%", padding: 2 }}
            {...{ onClick, onMouseDown, onMouseUp, onMouseEnter, onMouseLeave }}
        >
            <Cursor width={width - 4} height={height - 4} />
        </div>
    )
}

Cursors.defaultProps = {
    height: 45,
    width: 45,
    platform: "MacOS",
}

addPropertyControls(Cursors, {
    // platform: {
    //     title: "Platform",
    //     type: ControlType.Enum,
    //     options: ["MacOS"],
    //     defaultValue: Cursors.defaultProps.platform,
    // },
    macOSCursor: {
        title: "Cursor",
        type: ControlType.Enum,
        options: Object.keys(MacOSCursors),
        defaultValue: "Pointer",
    },
    ...defaultEvents,
})
