// @framer/framer.default/code/Input.tsx
import * as React from "react"
import { ControlType, addPropertyControls, RenderTarget } from "framer"
import { useState, useEffect, useRef } from "react"
import { fontStack, fontControls, fontSizeOptions } from "./lib/constants"
import { useUniqueClassName } from "./lib/useUniqueClassName"
import { useOnEnter } from "./lib/useOnNavigationTargetChange"

/*
 ** INPUT
 ** - Validation
 ** - Show Password button
 */

interface Props {
    value: string
    onValueChange: (value: string) => void
    placeholder: string
    backgroundColor: string
    textColor: string
    focusColor: string
    fontSize: number
    width: number
    height: number
    radius: number
    padding: number
    paddingPerSide: boolean
    paddingTop: number
    paddingRight: number
    paddingBottom: number
    paddingLeft: number
    border: string
    borderWidth: number
    multiLine: boolean
    password: boolean
    onChange: any
    style?: React.CSSProperties
    fontFamily?: string
    blurOnSubmit?: boolean
}

export function Input(props) {
    const {
        placeholder,
        backgroundColor,
        textColor,
        fontSize,
        fontWeight,
        radius,
        paddingPerSide,
        padding,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft,
        border,
        borderWidth,
        password,
        onSubmit,
        onFocus,
        onBlur,
        value,
        topLeft,
        topRight,
        bottomRight,
        bottomLeft,
        isMixed,
        multiLine,
        placeholderColor,
        focused,
        inputStyle,
        fontFamily,
        blurOnSubmit,
        keyboard,
    } = props

    const [inputValue, setValue] = useState(value)
    const inputEle = useRef<HTMLInputElement | HTMLTextAreaElement>()
    const className = useUniqueClassName("input", [placeholderColor])
    const inPreview = RenderTarget.current() === RenderTarget.preview
    const Tag = multiLine ? "textarea" : "input"
    const fontFamilyStyle = fontFamily ? { fontFamily } : {}

    const paddingValue = paddingPerSide
        ? `${paddingTop}px ${paddingRight}px ${paddingBottom}px ${paddingLeft}px`
        : padding
    const borderRadius = isMixed ? `${topLeft}px ${topRight}px ${bottomRight}px ${bottomLeft}px` : `${radius}px`

    const onChange = (event: React.ChangeEvent) => {
        const element = multiLine
            ? (event.nativeEvent.target as HTMLTextAreaElement)
            : (event.nativeEvent.target as HTMLInputElement)

        const value = element.value

        if (props.onChange) props.onChange(value)

        setValue(value)

        if (props.onValueChange) {
            props.onValueChange(value)
        }
    }

    React.useEffect(() => {
        setValue(value)
    }, [value])

    useOnEnter(() => {
        if (inPreview && focused) inputEle.current.focus()
    })

    return (
        <>
            <style>
                {`        
                    .${className}:focus {
                        box-shadow: inset 0 0 0 ${props.borderWidth}px
                        ${props.focusColor} !important;
                    }

                    .${className}::placeholder {
                        color: ${placeholderColor} !important;
                    }
                `}
            </style>
            <Tag
                onChange={onChange}
                ref={inputEle as any}
                value={inputValue}
                placeholder={placeholder}
                onKeyDown={e => {
                    if (e.keyCode === 13) {
                        // @ts-ignore
                        if (blurOnSubmit && inputEle.current) inputEle.current.blur()
                        if (onSubmit) onSubmit()
                    }
                }}
                onFocus={() => {
                    if (onFocus) onFocus()
                }}
                onBlur={() => {
                    if (onBlur) onBlur()
                }}
                autoFocus={inPreview && focused}
                className={className}
                style={{
                    ...style,
                    backgroundColor,
                    color: textColor,
                    fontSize: fontSize,
                    fontWeight: fontWeight,
                    borderRadius,
                    padding: paddingValue,
                    boxShadow:
                        !inPreview && focused
                            ? `inset 0 0 0 ${props.borderWidth}px ${props.focusColor}`
                            : `inset 0 0 0 ${borderWidth}px ${border}`,
                    ...fontFamilyStyle,
                    ...inputStyle,
                }}
                type={password ? "password" : "text"}
                inputMode={keyboard}
            />
        </>
    )
}

Input.defaultProps = {
    value: "",
    placeholder: "Type something…",
    width: 260,
    height: 50,
    backgroundColor: "#EBEBEB",
    textColor: "#333",
    focusColor: "#09F",
    fontSize: 16,
    fontWeight: 400,
    radius: 8,
    padding: 15,
    paddingPerSide: false,
    paddingTop: 12,
    paddingRight: 12,
    paddingBottom: 12,
    paddingLeft: 12,
    border: "rgba(0,0,0,0)",
    placeholderColor: "#aaa",
    borderWidth: 1,
    multiLine: false,
    password: false,
    keyboard: "",
}

addPropertyControls(Input, {
    placeholder: { type: ControlType.String, title: "Placeholder" },
    value: { type: ControlType.String, title: "Value" },
    textColor: { type: ControlType.Color, title: "Text Color" },
    placeholderColor: { type: ControlType.Color, title: "Placeholder" },
    backgroundColor: { type: ControlType.Color, title: "Background" },
    border: { type: ControlType.Color, title: "Border" },
    borderWidth: {
        type: ControlType.Number,
        title: " ",
        min: 1,
        max: 5,
        displayStepper: true,
    },

    focusColor: { type: ControlType.Color, title: "Focus" },
    focused: {
        type: ControlType.Boolean,
        title: "Focused",
        defaultValue: false,
        disabledTitle: "No",
        enabledTitle: "Yes",
    },
    ...fontControls,
    fontSize: {
        ...(fontSizeOptions as any),
    },
    padding: {
        type: ControlType.FusedNumber,
        toggleKey: "paddingPerSide",
        toggleTitles: ["Padding", "Padding per side"],
        valueKeys: ["paddingTop", "paddingRight", "paddingBottom", "paddingLeft"],
        valueLabels: ["T", "R", "B", "L"],
        min: 0,
        title: "Padding",
    },
    radius: {
        title: "Radius",
        type: ControlType.FusedNumber,
        defaultValue: Input.defaultProps.radius,
        toggleKey: "isMixed",
        toggleTitles: ["Radius", "Radius per corner"],
        valueKeys: ["topLeft", "topRight", "bottomRight", "bottomLeft"],
        valueLabels: ["TL", "TR", "BR", "BL"],
        min: 0,
    },

    multiLine: {
        type: ControlType.Boolean,
        title: "Text Area",
        defaultValue: false,
        disabledTitle: "No",
        enabledTitle: "Yes",
    },
    password: {
        type: ControlType.Boolean,
        title: "Password",
        hidden: ({ multiLine }) => multiLine,
        disabledTitle: "No",
        enabledTitle: "Yes",
    },
    keyboard: {
        type: ControlType.Enum,
        title: "Keyboard",
        defaultValue: "",
        options: ["", "numeric", "tel", "decimal", "email", "url", "search"],
        optionTitles: ["Default", "Numeric", "Phone", "Decimal", "Email", "URL", "Search"],
    },
    onChange: { type: ControlType.EventHandler },
    onSubmit: { type: ControlType.EventHandler },
    onFocus: { type: ControlType.EventHandler },
    onBlur: { type: ControlType.EventHandler },
})

const style: React.CSSProperties = {
    border: "none",
    width: "100%",
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    outline: "none",
    resize: "none",
    margin: 0,
    fontFamily: fontStack,
    WebkitTapHighlightColor: "rgba(0, 0, 0, 0)",
    WebkitAppearance: "none",
}
