// @framer/framer.default/code/Loading.tsx
import * as React from "react"
import { addPropertyControls, ControlType, useAnimation, Stack, motion } from "framer"
import { indentTitle } from "./lib/indentTitle"
import { defaultEvents } from "./lib/constants"
import { useOnEnter, useOnExit } from "./lib/useOnNavigationTargetChange"
import { useRef, useEffect } from "react"

enum Indicators {
    DotWave = "Dots",
    // Spinner = "Spinner",
    Material = "Material",
    IOS = "iOS",
}

const angleInRadians = (angleInDegrees) => (angleInDegrees - 90) * (Math.PI / 180.0)

const polarToCartesian = (centerX, centerY, radius, angleInDegrees) => {
    const a = angleInRadians(angleInDegrees)
    return {
        x: centerX + radius * Math.cos(a),
        y: centerY + radius * Math.sin(a),
    }
}

const arc = (x, y, radius, startAngle, endAngle) => {
    const fullCircle = endAngle - startAngle === 360
    const start = polarToCartesian(x, y, radius, endAngle - 0.01)
    const end = polarToCartesian(x, y, radius, startAngle)
    const arcFlag = endAngle - startAngle <= 180 ? "0" : "1"

    let d = ["M", start.x, start.y, "A", radius, radius, 0, arcFlag, 0, end.x, end.y].join(" ")

    if (fullCircle) d += "z"
    return d
}

function Spinner({ color }) {
    const length = 360
    const endPercentage = (length / 360) * 100
    const strokeWidth = 10
    const width = 100
    const height = 100

    return (
        <motion.div
            style={{
                height: "85%",
                width: "85%",
                position: "relative",
                originX: 0.5,
                originY: 0.5,
            }}
            animate={{ rotate: 360 }}
            transition={{ loop: Infinity, ease: "linear", duration: 0.5 }}
        >
            <motion.svg
                style={{
                    height: "100%",
                    width: "100%",
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0,

                    position: "absolute",
                    WebkitMask: `conic-gradient(rgba(0, 0, 0, 0.0) 0%, rgba(0, 0, 0,1.0) ${endPercentage}%)`,
                }}
                viewBox="0 0 100 100"
            >
                <g transform="translate(0 0)">
                    <path
                        d={arc(width / 2, height / 2, width / 2 - strokeWidth / 2, 0, length)}
                        fill={"none"}
                        stroke={color}
                        strokeWidth={strokeWidth}
                        strokeLinecap={"round"}
                    />
                </g>
            </motion.svg>
            <motion.svg
                style={{
                    height: "100%",
                    width: "100%",
                    position: "absolute",
                    // border: "1px solid red",
                }}
                viewBox="0 0 100 100"
            >
                <g transform="translate(50 0)">
                    <path d="M 0 0 C 2.761 0 5 2.239 5 5 C 5 7.761 2.761 10 0 10 C 0 10 0 0 0 0 Z" fill={color}></path>
                </g>
            </motion.svg>
        </motion.div>
    )
}
// <path d="M 0 0 C 2.761 0 5 2.239 5 5 C 5 7.761 2.761 10 0 10 C 0 10 0 0 0 0 Z" fill="#CCC"></path>

// function Spinner({ color }) {
//     return (
//         <svg style={{ height: "85%", width: "85%" }} viewBox="0 0 100 100">
//             <motion.g
//                 transform="translate(3 3)"
//                 animate={{ rotate: 360 }}
//                 transition={{ loop: Infinity, ease: "linear", duration: 1 }}
//             >
//                 {pathStrings.map((data, i) => {
//                     return <path d={data} fill={color} opacity={i / pathStrings.length} />
//                 })}
//             </motion.g>
//         </svg>
//     )
// }

function DotWave({ color }) {
    const circles = [0, 1, 2]

    return (
        <motion.svg
            style={{
                height: "85%",
                width: "85%",
            }}
            viewBox="0 0 30 30"
            variants={{
                show: {
                    transition: {
                        delayChildren: 0.1,
                        staggerChildren: 0.12,
                    },
                },
            }}
            animate={"show"}
        >
            {circles.map((circle) => (
                <motion.circle
                    style={{ fill: color }}
                    variants={{
                        hidden: { y: 0 },
                        show: { y: [0, 0, 0, -10, 0, 0, 0] },
                    }}
                    key={circle}
                    transition={{
                        yoyo: Infinity,
                        duration: 1.45,
                        ease: "easeInOut",
                    }}
                    r={3}
                    cx={circle * 10 + 5}
                    cy={15}
                />
            ))}
        </motion.svg>
    )
}

function Material({ color }) {
    return (
        <motion.svg
            style={{
                height: "85%",
                width: "85%",
                overflow: "visible",
                originX: "50%",
                originY: "50%",
            }}
            animate={{ rotate: 360 }}
            transition={{ ease: "linear", loop: Infinity, duration: 2 }}
            viewBox="25 25 50 50"
        >
            <motion.circle
                style={{
                    stroke: color,
                    strokeLinecap: "round",
                }}
                animate={{
                    strokeDasharray: ["1, 200", "89, 200", "89, 200"],
                    strokeDashoffset: [0, -35, -124],
                }}
                transition={{
                    duration: 1.5,
                    loop: Infinity,
                    ease: "easeInOut",
                }}
                cx="50"
                cy="50"
                r="20"
                fill="none"
                strokeWidth={2}
                strokeMiterlimit="10"
            />
        </motion.svg>
    )
}

function IOS({ color }) {
    const particles = 12
    const duration = 1
    // this was the death of me
    const arrayRotate = (arr, n) => arr.slice(n, arr.length).concat(arr.slice(0, n))
    const lines = [...new Array(particles)].map((l, i) => (0.9 / particles) * i + 0.1).reverse()
    const lineOpacities = lines.map((l, i) => arrayRotate(lines, i))

    return (
        <motion.svg viewBox="-15 -15 30 30" style={{ width: "100%", height: "100%" }}>
            {lineOpacities.map((lineKeyframes, i) => (
                <motion.g
                    key={i}
                    initial={{ opacity: lineKeyframes[0] }}
                    animate={{ opacity: lineKeyframes }}
                    transition={{
                        duration,
                        loop: Infinity,
                        repeatDelay: 0.0005,
                    }}
                >
                    <rect
                        style={{
                            width: 7,
                            height: 2,
                            fill: color,
                            transform: `rotate(${((particles - i) / particles) * 360 - 90}deg)`,
                        }}
                        x={5}
                        y={-1}
                        rx={1}
                    />
                </motion.g>
            ))}
        </motion.svg>
    )
}

function getIndicator(indicator, props) {
    switch (indicator) {
        case Indicators.DotWave:
            return <DotWave {...props} />
        case Indicators.Material:
            return <Material {...props} />
        case Indicators.IOS:
            return <IOS {...props} />
        // case Indicators.Spinner:
        //     return <Spinner {...props} />
        default:
            return <DotWave {...props} />
    }
}

export function handleTimeout(duration: number, callback: () => void) {
    const id = setTimeout(callback, duration * 1000)
    return () => clearTimeout(id)
}

export function Loading(props) {
    const {
        duration,
        onTimeout,
        fadeOut,
        hasDuration,
        indicator,
        onClick,
        onMouseDown,
        onMouseUp,
        onMouseEnter,
        onMouseLeave,
    } = props

    const controls = useAnimation()
    const animDuration = fadeOut ? Math.min(duration, 0.35) : 0
    const animDelay = fadeOut ? duration - animDuration : duration
    const currentIndicator = getIndicator(indicator, props)
    const handlers = useRef<(() => void)[]>([])

    const onFadeOut = React.useCallback(() => {
        if (hasDuration)
            controls.start({
                opacity: 0,
                transition: { duration: animDuration, ease: "easeIn" },
            })
    }, [hasDuration, animDuration])

    const resetOpacity = async () => {
        controls.set({ opacity: 1 })
    }

    useOnEnter(() => {
        resetOpacity()
        if (hasDuration) handlers.current = [handleTimeout(duration, onTimeout), handleTimeout(animDelay, onFadeOut)]
    })

    // Cancel all timers on exit.
    useOnExit(() => handlers.current.forEach((cleanup) => cleanup))

    // Cancel all timers on unmount.
    useEffect(() => () => handlers.current.forEach((cleanup) => cleanup), [])

    return (
        <motion.div
            {...{ onClick, onMouseDown, onMouseUp, onMouseEnter, onMouseLeave }}
            animate={controls}
            style={{
                position: "relative",
                overflow: "show",
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                width: "100%",
                height: "100%",
            }}
        >
            {currentIndicator}
        </motion.div>
    )
}

Loading.defaultProps = {
    height: 40,
    width: 40,
    duration: 2,
    hasDuration: false,
}

// Learn more: https://framer.com/api/property-controls/
addPropertyControls(Loading, {
    indicator: {
        title: "Indicator",
        type: ControlType.Enum,
        options: Object.keys(Indicators).map((i) => Indicators[i]),
    },
    color: {
        type: ControlType.Color,
        defaultValue: "#888",
    },
    // transition: { title: "Animation", type: ControlType.Transition },
    hasDuration: {
        title: "Duration",
        type: ControlType.Boolean,
        defaultValue: Loading.defaultProps.hasDuration,
        enabledTitle: "Timeout",
        disabledTitle: "Infinity",
    },
    duration: {
        title: indentTitle("Time"),
        hidden: ({ hasDuration }) => !hasDuration,
        min: 0.1,
        max: 10,
        defaultValue: Loading.defaultProps.duration,
        type: ControlType.Number,
        step: 0.1,
    },
    fadeOut: {
        title: indentTitle("Fade Out"),
        hidden: ({ hasDuration }) => !hasDuration,
        type: ControlType.Boolean,
        enabledTitle: "Yes",
        disabledTitle: "No",
    },
    onTimeout: {
        type: ControlType.EventHandler,
    },
    ...defaultEvents,
})

const pathStrings = [
    "M 48.25 0.03 L 49.79 0.08 L 50.08 0.1 L 50.37 0.12 L 50.66 0.14 L 50.95 0.16 L 51.41 0.21 L 51.7 0.24 L 51.99 0.27 L 52.28 0.3 L 51.26 10.24 L 50.97 10.21 L 50.68 10.18 L 50.39 10.15 L 50.27 10.14 L 49.98 10.12 L 49.69 10.1 L 49.4 10.08 L 49.11 10.06 L 47.91 10.02 Z",
    "M 52.28 0.3 L 52.57 0.33 L 53.03 0.39 L 53.15 0.4 L 53.63 0.46 L 54.06 0.53 L 54.18 0.55 L 54.64 0.62 L 54.78 0.64 L 55.21 0.71 L 53.51 10.57 L 53.36 10.54 L 52.94 10.48 L 52.82 10.45 L 52.36 10.39 L 52.21 10.36 L 52.13 10.34 L 51.67 10.29 L 51.55 10.27 L 51.26 10.24 Z",
    "M 55.21 0.71 L 55.5 0.76 L 55.98 0.86 L 56.07 0.87 L 56.55 0.97 L 56.8 1.02 L 57.12 1.09 L 57.53 1.19 L 57.69 1.22 L 58.13 1.33 L 55.71 11.03 L 55.59 11 L 55.19 10.91 L 55.02 10.87 L 54.78 10.82 L 54.45 10.75 L 54.37 10.73 L 53.88 10.64 L 53.8 10.62 L 53.51 10.57 Z",
    "M 58.13 1.33 L 58.41 1.4 L 58.66 1.46 L 58.98 1.54 L 59.42 1.66 L 59.7 1.74 L 59.82 1.77 L 60.42 1.95 L 60.54 1.98 L 60.91 2.1 L 57.97 11.64 L 57.8 11.6 L 57.36 11.47 L 57.4 11.47 L 56.96 11.36 L 56.68 11.28 L 56.56 11.24 L 56.32 11.18 L 55.99 11.1 L 55.71 11.03 Z",
    "M 60.91 2.1 L 61.25 2.2 L 61.53 2.29 L 61.81 2.38 L 62.13 2.49 L 62.51 2.62 L 62.84 2.74 L 62.95 2.78 L 63.33 2.91 L 63.73 3.07 L 60.11 12.39 L 59.97 12.33 L 59.79 12.26 L 59.36 12.12 L 59.15 12.04 L 58.97 11.97 L 58.75 11.9 L 58.47 11.81 L 58.19 11.72 L 57.97 11.64 Z",
    "M 63.73 3.07 L 63.93 3.14 L 64.35 3.31 L 64.62 3.42 L 64.89 3.53 L 65.16 3.64 L 65.43 3.75 L 65.9 3.96 L 65.96 3.98 L 66.41 4.18 L 62.28 13.29 L 62.18 13.24 L 61.7 13.04 L 61.65 13.01 L 61.38 12.9 L 61.11 12.79 L 60.84 12.68 L 60.57 12.57 L 60.45 12.52 L 60.11 12.39 Z",
    "M 66.41 4.18 L 66.7 4.31 L 67.1 4.51 L 67.16 4.53 L 67.63 4.76 L 67.89 4.89 L 68.15 5.02 L 68.41 5.15 L 68.67 5.28 L 69.09 5.51 L 64.28 14.27 L 64.19 14.22 L 63.93 14.09 L 63.67 13.96 L 63.41 13.83 L 63.15 13.7 L 63.1 13.67 L 62.62 13.45 L 62.5 13.39 L 62.28 13.29 Z",
    "M 69.09 5.51 L 69.31 5.63 L 69.57 5.77 L 69.89 5.95 L 70.14 6.09 L 70.33 6.19 L 70.78 6.45 L 71.03 6.6 L 71.28 6.75 L 71.57 6.93 L 66.35 15.45 L 66.14 15.33 L 65.89 15.18 L 65.64 15.03 L 65.59 14.99 L 65.26 14.81 L 65.01 14.67 L 64.83 14.57 L 64.57 14.43 L 64.28 14.27 Z",
    "M 71.57 6.93 L 71.85 7.1 L 72.15 7.29 L 72.34 7.41 L 72.64 7.6 L 72.95 7.81 L 73.19 7.97 L 73.55 8.22 L 73.67 8.3 L 74.03 8.55 L 68.25 16.71 L 68.13 16.62 L 67.77 16.38 L 67.65 16.29 L 67.41 16.13 L 67.24 16.02 L 67.04 15.89 L 66.75 15.71 L 66.55 15.58 L 66.35 15.45 Z",
    "M 74.03 8.55 L 74.27 8.72 L 74.51 8.89 L 74.82 9.12 L 75.05 9.29 L 75.21 9.4 L 75.63 9.72 L 75.86 9.9 L 76.09 10.08 L 76.36 10.29 L 70.11 18.11 L 69.93 17.96 L 69.7 17.78 L 69.47 17.6 L 69.43 17.56 L 69.11 17.33 L 68.88 17.16 L 68.73 17.05 L 68.49 16.88 L 68.25 16.71 Z",
    "M 76.36 10.29 L 76.54 10.44 L 76.87 10.72 L 77.08 10.88 L 77.4 11.16 L 77.62 11.35 L 77.84 11.54 L 78.06 11.73 L 78.28 11.92 L 78.59 12.2 L 71.88 19.61 L 71.74 19.48 L 71.52 19.29 L 71.3 19.1 L 71.08 18.91 L 70.86 18.72 L 70.74 18.62 L 70.51 18.42 L 70.38 18.32 L 70.11 18.11 Z",
    "M 78.59 12.2 L 78.79 12.38 L 79.1 12.67 L 79.23 12.79 L 79.53 13.07 L 79.82 13.35 L 80.03 13.56 L 80.15 13.68 L 80.52 14.06 L 80.69 14.23 L 73.53 21.22 L 73.28 20.96 L 73.25 20.92 L 72.95 20.64 L 72.74 20.43 L 72.63 20.31 L 72.51 20.19 L 72.2 19.91 L 72.09 19.8 L 71.88 19.61 Z",
    "M 80.69 14.23 L 80.93 14.48 L 81.13 14.69 L 81.33 14.9 L 81.53 15.11 L 81.88 15.51 L 81.92 15.54 L 82.27 15.94 L 82.46 16.16 L 82.65 16.38 L 75.09 22.92 L 74.9 22.7 L 74.71 22.48 L 74.68 22.44 L 74.32 22.05 L 74.29 22.01 L 74.09 21.8 L 73.89 21.59 L 73.69 21.38 L 73.53 21.22 Z",
    "M 82.65 16.38 L 82.84 16.6 L 83.03 16.82 L 83.29 17.14 L 83.49 17.37 L 83.74 17.69 L 83.92 17.92 L 84.03 18.05 L 84.28 18.37 L 84.53 18.7 L 76.51 24.68 L 76.4 24.53 L 76.29 24.39 L 76.04 24.08 L 75.86 23.85 L 75.75 23.71 L 75.59 23.5 L 75.47 23.36 L 75.28 23.14 L 75.09 22.92 Z",
    "M 84.53 18.7 L 84.71 18.95 L 84.88 19.18 L 85.04 19.39 L 85.23 19.65 L 85.53 20.09 L 85.62 20.21 L 85.79 20.45 L 86.03 20.81 L 86.19 21.05 L 77.87 26.59 L 77.71 26.35 L 77.63 26.23 L 77.46 25.99 L 77.21 25.63 L 77.19 25.59 L 77.04 25.39 L 76.84 25.12 L 76.67 24.89 L 76.51 24.68 Z",
    "M 86.19 21.05 L 86.35 21.29 L 86.56 21.61 L 86.74 21.9 L 86.87 22.1 L 87.1 22.48 L 87.25 22.73 L 87.35 22.89 L 87.59 23.3 L 87.73 23.55 L 79.09 28.55 L 78.93 28.3 L 78.87 28.19 L 78.67 27.87 L 78.52 27.62 L 78.45 27.5 L 78.26 27.2 L 78.14 27.01 L 78.03 26.83 L 77.87 26.59 Z",
    "M 87.73 23.55 L 87.91 23.86 L 88.05 24.11 L 88.17 24.31 L 88.41 24.75 L 88.51 24.95 L 88.65 25.21 L 88.82 25.52 L 88.98 25.85 L 89.11 26.11 L 80.17 30.59 L 80.04 30.33 L 79.94 30.14 L 79.85 29.95 L 79.71 29.69 L 79.53 29.37 L 79.51 29.31 L 79.33 28.99 L 79.19 28.74 L 79.09 28.55 Z",
    "M 89.11 26.11 L 89.24 26.37 L 89.41 26.71 L 89.56 27.04 L 89.62 27.16 L 89.84 27.64 L 89.93 27.83 L 90.13 28.31 L 90.19 28.44 L 90.31 28.74 L 81.14 32.72 L 81.05 32.5 L 80.87 32.09 L 80.85 32.03 L 80.7 31.7 L 80.68 31.64 L 80.48 31.24 L 80.39 31.05 L 80.3 30.85 L 80.17 30.59 Z",
    "M 90.31 28.74 L 90.42 28.97 L 90.59 29.38 L 90.75 29.8 L 90.8 29.92 L 90.96 30.34 L 91.03 30.53 L 91.17 30.89 L 91.27 31.16 L 91.4 31.55 L 81.94 34.81 L 81.89 34.64 L 81.79 34.37 L 81.73 34.19 L 81.58 33.82 L 81.54 33.7 L 81.37 33.28 L 81.33 33.16 L 81.28 33.03 L 81.14 32.72 Z",
    "M 91.4 31.55 L 91.46 31.71 L 91.62 32.2 L 91.69 32.42 L 91.8 32.75 L 91.89 33.03 L 92.02 33.47 L 92.08 33.69 L 92.14 33.86 L 92.29 34.38 L 82.63 36.96 L 82.62 36.92 L 82.5 36.53 L 82.4 36.21 L 82.37 36.09 L 82.28 35.81 L 82.21 35.58 L 82.1 35.26 L 82.08 35.19 L 81.94 34.81 Z",
    "M 92.29 34.38 L 92.38 34.74 L 92.42 34.86 L 92.54 35.35 L 92.6 35.59 L 92.67 35.87 L 92.74 36.15 L 92.85 36.64 L 92.87 36.72 L 92.97 37.19 L 83.19 39.25 L 83.17 39.14 L 83.05 38.66 L 83.04 38.57 L 82.97 38.29 L 82.9 38.01 L 82.82 37.69 L 82.8 37.6 L 82.68 37.16 L 82.63 36.96 Z",
    "M 92.97 37.19 L 93.03 37.45 L 93.09 37.73 L 93.18 38.22 L 93.21 38.35 L 93.28 38.76 L 93.34 39.08 L 93.4 39.54 L 93.42 39.62 L 93.48 40.02 L 83.6 41.55 L 83.58 41.38 L 83.5 40.9 L 83.48 40.78 L 83.44 40.52 L 83.41 40.37 L 83.32 39.92 L 83.31 39.83 L 83.25 39.55 L 83.19 39.25 Z",
    "M 93.48 40.02 L 93.52 40.23 L 93.58 40.69 L 93.63 41.15 L 93.65 41.23 L 93.7 41.72 L 93.73 42.01 L 93.76 42.3 L 93.79 42.59 L 93.83 42.97 L 83.87 43.81 L 83.85 43.61 L 83.82 43.32 L 83.79 43.03 L 83.76 42.74 L 83.75 42.65 L 83.69 42.17 L 83.68 42.05 L 83.66 41.93 L 83.6 41.55 Z",
    "M 93.83 42.97 L 93.86 43.34 L 93.88 43.63 L 93.9 43.92 L 93.92 44.21 L 93.94 44.5 L 93.96 44.96 L 93.97 45.25 L 93.98 45.54 L 93.99 45.91 L 83.99 46.09 L 83.98 45.88 L 83.97 45.59 L 83.96 45.3 L 83.96 45.18 L 83.94 44.89 L 83.92 44.6 L 83.9 44.31 L 83.88 44.02 L 83.87 43.81 Z",
    "M 93.99 45.91 L 93.99 46.29 L 94 46.42 L 94 47.46 L 93.99 47.92 L 93.99 48.04 L 93.98 48.5 L 93.97 48.79 L 83.97 48.45 L 83.98 48.16 L 83.99 48.04 L 83.99 47.58 L 84 47.46 L 84 46.76 L 83.99 46.29 L 83.99 46.09 Z",
    "M 93.97 48.79 L 93.96 49.08 L 93.93 49.54 L 93.93 49.66 L 93.9 50.12 L 93.88 50.41 L 93.86 50.7 L 93.81 51.16 L 93.81 51.28 L 93.76 51.74 L 83.82 50.72 L 83.83 50.6 L 83.87 50.14 L 83.88 50.02 L 83.9 49.73 L 83.92 49.44 L 83.93 49.32 L 83.95 48.86 L 83.96 48.74 L 83.97 48.45 Z",
    "M 93.76 51.74 L 93.73 52.03 L 93.7 52.32 L 93.64 52.78 L 93.63 52.9 L 93.57 53.36 L 93.53 53.67 L 93.49 53.93 L 93.42 54.39 L 93.39 54.61 L 83.52 53.05 L 83.56 52.69 L 83.59 52.57 L 83.63 52.25 L 83.67 52 L 83.69 51.88 L 83.74 51.42 L 83.76 51.3 L 83.79 51.01 L 83.82 50.72 Z",
    "M 93.39 54.61 L 93.32 54.99 L 93.28 55.25 L 93.23 55.54 L 93.13 56.02 L 93.12 56.11 L 93.02 56.59 L 92.97 56.84 L 92.9 57.16 L 92.82 57.51 L 83.07 55.29 L 83.12 55.06 L 83.17 54.82 L 83.24 54.49 L 83.26 54.41 L 83.35 53.92 L 83.37 53.84 L 83.42 53.55 L 83.48 53.23 L 83.52 53.05 Z",
    "M 92.82 57.51 L 92.73 57.89 L 92.66 58.17 L 92.59 58.45 L 92.53 58.7 L 92.45 59.02 L 92.33 59.46 L 92.25 59.74 L 92.17 60.02 L 92.09 60.3 L 82.47 57.56 L 82.55 57.28 L 82.63 57 L 82.71 56.72 L 82.75 56.6 L 82.81 56.36 L 82.89 56.03 L 82.96 55.75 L 83.03 55.47 L 83.07 55.29 Z",
    "M 92.09 60.3 L 92.01 60.58 L 91.85 61.06 L 91.84 61.13 L 91.7 61.57 L 91.59 61.89 L 91.46 62.27 L 91.42 62.4 L 91.25 62.88 L 91.16 63.12 L 81.76 59.7 L 81.87 59.4 L 81.9 59.34 L 82.04 58.91 L 82.11 58.73 L 82.18 58.51 L 82.22 58.39 L 82.37 57.9 L 82.39 57.84 L 82.47 57.56 Z",
    "M 91.16 63.12 L 91.07 63.37 L 90.95 63.7 L 90.78 64.12 L 90.74 64.24 L 90.57 64.66 L 90.46 64.93 L 90.35 65.2 L 90.17 65.61 L 90.08 65.83 L 80.91 61.85 L 81.03 61.55 L 81.09 61.42 L 81.2 61.15 L 81.31 60.88 L 81.36 60.76 L 81.52 60.34 L 81.57 60.22 L 81.65 60.01 L 81.76 59.7 Z",
    "M 90.08 65.83 L 89.94 66.14 L 89.79 66.47 L 89.7 66.67 L 89.55 67 L 89.35 67.4 L 89.26 67.6 L 89.09 67.93 L 88.96 68.19 L 88.81 68.48 L 79.91 63.94 L 80.02 63.71 L 80.15 63.45 L 80.24 63.26 L 80.41 62.92 L 80.47 62.8 L 80.56 62.61 L 80.71 62.27 L 80.8 62.08 L 80.91 61.85 Z",
    "M 88.81 68.48 L 88.67 68.77 L 88.49 69.09 L 88.35 69.35 L 88.17 69.67 L 88.07 69.86 L 87.89 70.18 L 87.75 70.43 L 87.53 70.81 L 87.38 71.06 L 78.8 65.92 L 78.95 65.67 L 79.03 65.55 L 79.17 65.3 L 79.27 65.12 L 79.45 64.79 L 79.55 64.61 L 79.69 64.35 L 79.79 64.15 L 79.91 63.94 Z",
    "M 87.38 71.06 L 87.23 71.31 L 87.08 71.56 L 86.93 71.81 L 86.64 72.25 L 86.53 72.43 L 86.41 72.62 L 86.22 72.92 L 85.92 73.35 L 85.84 73.47 L 77.52 67.93 L 77.76 67.57 L 77.8 67.52 L 77.93 67.32 L 78.11 67.03 L 78.32 66.71 L 78.35 66.67 L 78.5 66.42 L 78.65 66.17 L 78.8 65.92 Z",
    "M 85.84 73.47 L 85.68 73.71 L 85.43 74.07 L 85.2 74.38 L 85.09 74.54 L 84.86 74.85 L 84.75 75.01 L 84.43 75.43 L 84.25 75.66 L 84.12 75.84 L 76.16 69.78 L 76.37 69.5 L 76.55 69.27 L 76.59 69.23 L 76.82 68.91 L 76.93 68.76 L 77.16 68.44 L 77.27 68.29 L 77.36 68.17 L 77.52 67.93 Z",
    "M 84.12 75.84 L 83.9 76.12 L 83.62 76.45 L 83.46 76.66 L 83.35 76.8 L 83 77.21 L 82.81 77.43 L 82.79 77.47 L 82.27 78.05 L 82.24 78.09 L 74.68 71.55 L 75.03 71.15 L 74.91 71.31 L 75.25 70.89 L 75.44 70.67 L 75.47 70.64 L 75.72 70.32 L 75.92 70.09 L 76.02 69.96 L 76.16 69.78 Z",
    "M 82.24 78.09 L 82.05 78.31 L 81.77 78.62 L 81.59 78.82 L 81.3 79.13 L 81.18 79.26 L 80.9 79.56 L 80.62 79.85 L 80.41 80.06 L 80.21 80.27 L 73.13 73.19 L 73.33 72.98 L 73.54 72.77 L 73.66 72.66 L 73.78 72.54 L 74.06 72.23 L 74.17 72.12 L 74.37 71.9 L 74.49 71.77 L 74.68 71.55 Z",
    "M 80.21 80.27 L 79.91 80.55 L 79.79 80.68 L 79.49 80.96 L 79.28 81.16 L 79.07 81.36 L 78.86 81.56 L 78.46 81.91 L 78.43 81.95 L 78.03 82.3 L 71.49 74.74 L 71.53 74.71 L 71.92 74.35 L 71.96 74.32 L 72.17 74.12 L 72.38 73.92 L 72.59 73.72 L 72.71 73.6 L 73.01 73.31 L 73.13 73.19 Z",
    "M 78.03 82.3 L 77.81 82.49 L 77.59 82.68 L 77.37 82.87 L 77.15 83.06 L 76.73 83.4 L 76.7 83.43 L 76.28 83.77 L 76.05 83.95 L 75.82 84.13 L 69.66 76.25 L 69.89 76.07 L 70.12 75.89 L 70.16 75.87 L 70.57 75.52 L 70.61 75.5 L 70.83 75.31 L 71.05 75.12 L 71.27 74.93 L 71.49 74.74 Z",
    "M 75.82 84.13 L 75.59 84.31 L 75.25 84.56 L 75.13 84.66 L 74.7 84.97 L 74.55 85.08 L 74.23 85.31 L 74.08 85.42 L 73.64 85.72 L 73.46 85.85 L 67.8 77.61 L 68.1 77.4 L 68.14 77.38 L 68.45 77.15 L 68.61 77.04 L 68.92 76.81 L 68.97 76.78 L 69.31 76.52 L 69.43 76.43 L 69.66 76.25 Z",
    "M 73.46 85.85 L 73.16 86.05 L 72.84 86.26 L 72.67 86.37 L 72.43 86.53 L 72.11 86.74 L 71.82 86.92 L 71.49 87.12 L 71.37 87.2 L 70.92 87.45 L 65.91 78.81 L 65.97 78.78 L 66.35 78.54 L 66.52 78.44 L 66.71 78.32 L 66.89 78.21 L 67.13 78.05 L 67.44 77.84 L 67.62 77.73 L 67.8 77.61 Z",
    "M 70.92 87.45 L 70.74 87.57 L 70.49 87.72 L 70.11 87.93 L 69.91 88.05 L 69.6 88.22 L 69.14 88.46 L 69.09 88.49 L 68.76 88.67 L 68.37 88.87 L 63.89 79.93 L 64.02 79.87 L 64.21 79.77 L 64.66 79.52 L 64.72 79.5 L 64.91 79.39 L 65.23 79.21 L 65.35 79.14 L 65.6 78.99 L 65.91 78.81 Z",
    "M 68.37 88.87 L 68.11 89 L 67.85 89.13 L 67.59 89.26 L 67.33 89.39 L 66.85 89.61 L 66.66 89.7 L 66.4 89.82 L 66.06 89.97 L 65.83 90.08 L 61.69 80.97 L 62 80.83 L 62.2 80.74 L 62.46 80.62 L 62.79 80.47 L 62.85 80.45 L 63.11 80.32 L 63.37 80.19 L 63.63 80.06 L 63.89 79.93 Z",
    "M 65.83 90.08 L 65.6 90.18 L 65.12 90.38 L 64.85 90.49 L 64.58 90.6 L 64.31 90.71 L 63.89 90.87 L 63.77 90.92 L 63.35 91.08 L 63.04 91.19 L 59.62 81.79 L 59.87 81.7 L 59.99 81.66 L 60.41 81.49 L 60.53 81.45 L 60.8 81.34 L 61.07 81.23 L 61.34 81.12 L 61.4 81.1 L 61.69 80.97 Z",
    "M 63.04 91.19 L 62.8 91.28 L 62.46 91.4 L 62.09 91.52 L 61.76 91.63 L 61.54 91.7 L 61.21 91.81 L 60.93 91.9 L 60.49 92.03 L 60.33 92.09 L 57.38 82.53 L 57.75 82.41 L 57.87 82.38 L 58.15 82.29 L 58.38 82.22 L 58.7 82.11 L 58.93 82.04 L 59.1 81.98 L 59.32 81.9 L 59.62 81.79 Z",
    "M 60.33 92.09 L 59.94 92.2 L 59.66 92.28 L 59.22 92.39 L 59.1 92.43 L 58.61 92.55 L 58.37 92.61 L 58.09 92.68 L 57.81 92.75 L 57.42 92.84 L 55.2 83.09 L 55.39 83.05 L 55.67 82.98 L 55.95 82.91 L 56.27 82.83 L 56.36 82.81 L 56.8 82.69 L 56.92 82.66 L 57.2 82.58 L 57.38 82.53 Z",
    "M 57.42 92.84 L 57.24 92.88 L 56.75 92.99 L 56.51 93.04 L 56.02 93.13 L 55.94 93.15 L 55.45 93.24 L 55.2 93.28 L 54.88 93.34 L 54.61 93.38 L 52.88 83.53 L 53.18 83.48 L 53.44 83.44 L 53.75 83.38 L 53.84 83.37 L 54.32 83.27 L 54.41 83.26 L 54.73 83.19 L 54.82 83.18 L 55.2 83.09 Z",
    "M 54.61 93.38 L 54.34 93.43 L 53.85 93.5 L 53.56 93.54 L 53.27 93.58 L 52.98 93.62 L 52.55 93.67 L 52.41 93.69 L 51.95 93.74 L 51.66 93.77 L 50.64 83.83 L 50.93 83.8 L 51.05 83.79 L 51.49 83.73 L 51.62 83.72 L 51.91 83.68 L 52.2 83.64 L 52.49 83.6 L 52.58 83.59 L 52.88 83.53 Z",
    "M 51.66 93.77 L 51.37 93.8 L 50.91 93.84 L 50.62 93.86 L 50.5 93.87 L 49.87 93.91 L 49.75 93.92 L 49.46 93.94 L 49 93.96 L 48.71 93.97 L 48.37 83.97 L 48.66 83.96 L 48.78 83.96 L 49.07 83.94 L 49.53 83.91 L 49.48 83.93 L 49.94 83.88 L 50.23 83.86 L 50.35 83.86 L 50.64 83.83 Z",
    "M 48.71 93.97 L 48.42 93.98 L 48.12 93.99 L 47.66 93.99 L 47.54 94 L 46.5 94 L 46.04 93.99 L 45.84 93.99 L 46.01 83.99 L 46.38 83.99 L 46.5 84 L 47.2 84 L 47.66 83.99 L 47.78 83.99 L 48.08 83.98 L 48.37 83.97 Z",
    "M 45.84 93.99 L 45.46 93.98 L 45.17 93.97 L 44.88 93.96 L 44.42 93.93 L 44.13 93.91 L 44.01 93.91 L 43.55 93.88 L 43.09 93.83 L 42.88 93.82 L 43.73 83.86 L 44.11 83.89 L 44.23 83.9 L 44.35 83.91 L 44.81 83.93 L 45.1 83.95 L 45.22 83.96 L 45.51 83.97 L 45.8 83.98 L 46.01 83.99 Z",
    "M 42.88 93.82 L 42.51 93.78 L 42.39 93.78 L 41.93 93.73 L 41.47 93.67 L 41.35 93.66 L 40.89 93.6 L 40.6 93.56 L 40.31 93.52 L 40.02 93.48 L 41.41 83.58 L 41.67 83.62 L 41.96 83.66 L 42.25 83.7 L 42.37 83.72 L 42.83 83.77 L 42.95 83.79 L 43.07 83.8 L 43.53 83.84 L 43.73 83.86 Z",
    "M 40.02 93.48 L 39.74 93.44 L 39.28 93.37 L 38.97 93.31 L 38.71 93.27 L 38.42 93.22 L 38.11 93.16 L 37.69 93.08 L 37.37 93.01 L 37.1 92.95 L 39.16 83.16 L 39.47 83.23 L 39.71 83.28 L 39.87 83.32 L 40.12 83.36 L 40.41 83.41 L 40.73 83.47 L 40.98 83.51 L 41.1 83.54 L 41.41 83.58 Z",
    "M 37.1 92.95 L 36.8 92.89 L 36.55 92.84 L 36.07 92.72 L 35.79 92.65 L 35.51 92.58 L 35.26 92.52 L 34.94 92.44 L 34.5 92.32 L 34.22 92.24 L 36.96 82.62 L 37.24 82.7 L 37.36 82.74 L 37.6 82.8 L 37.93 82.88 L 38.21 82.95 L 38.49 83.02 L 38.57 83.04 L 38.9 83.11 L 39.16 83.16 Z",
    "M 34.22 92.24 L 33.94 92.16 L 33.66 92.08 L 33.38 92 L 32.9 91.84 L 32.67 91.77 L 32.39 91.68 L 32.07 91.57 L 31.84 91.5 L 31.47 91.37 L 34.73 81.91 L 34.9 81.98 L 35.23 82.09 L 35.45 82.16 L 35.73 82.25 L 36.06 82.36 L 36.12 82.38 L 36.4 82.46 L 36.68 82.54 L 36.96 82.62 Z",
    "M 31.47 91.37 L 31.29 91.31 L 30.81 91.14 L 30.54 91.04 L 30.17 90.89 L 29.99 90.83 L 29.57 90.66 L 29.3 90.55 L 29.03 90.44 L 28.73 90.31 L 32.57 81.08 L 32.81 81.18 L 33.08 81.29 L 33.35 81.4 L 33.47 81.45 L 33.83 81.59 L 34.02 81.66 L 34.29 81.76 L 34.35 81.79 L 34.73 81.91 Z",
    "M 28.73 90.31 L 28.44 90.19 L 28.09 90.04 L 27.82 89.92 L 27.49 89.77 L 27.23 89.65 L 27.03 89.56 L 26.56 89.33 L 26.44 89.28 L 26.04 89.08 L 30.52 80.14 L 30.64 80.2 L 31.04 80.39 L 31.09 80.42 L 31.43 80.57 L 31.69 80.69 L 31.88 80.78 L 32.15 80.9 L 32.34 80.99 L 32.57 81.08 Z",
    "M 26.04 89.08 L 25.78 88.95 L 25.39 88.74 L 25.26 88.68 L 24.87 88.47 L 24.55 88.29 L 24.36 88.19 L 24.04 88.01 L 23.85 87.91 L 23.46 87.69 L 28.48 79.05 L 28.59 79.11 L 28.92 79.29 L 29.1 79.39 L 29.43 79.57 L 29.61 79.67 L 29.74 79.74 L 30.13 79.94 L 30.26 80.01 L 30.52 80.14 Z",
    "M 23.46 87.69 L 23.15 87.51 L 22.9 87.36 L 22.65 87.21 L 22.4 87.06 L 21.96 86.77 L 21.91 86.75 L 21.47 86.46 L 21.29 86.35 L 20.98 86.14 L 26.52 77.82 L 26.69 77.93 L 27.01 78.14 L 27.05 78.17 L 27.5 78.45 L 27.54 78.48 L 27.79 78.63 L 28.04 78.78 L 28.29 78.93 L 28.48 79.05 Z",
    "M 20.98 86.14 L 20.74 85.98 L 20.5 85.82 L 20.14 85.57 L 20.02 85.49 L 19.66 85.24 L 19.35 85.01 L 19.08 84.81 L 18.88 84.66 L 18.59 84.45 L 24.64 76.49 L 24.82 76.62 L 25.08 76.81 L 25.29 76.97 L 25.44 77.08 L 25.56 77.17 L 25.92 77.41 L 26.04 77.5 L 26.28 77.66 L 26.52 77.82 Z",
    "M 18.59 84.45 L 18.31 84.23 L 18.08 84.05 L 17.85 83.87 L 17.62 83.69 L 17.31 83.44 L 17.07 83.23 L 16.86 83.07 L 16.62 82.86 L 16.31 82.6 L 22.85 75.04 L 22.98 75.16 L 23.2 75.33 L 23.43 75.53 L 23.65 75.7 L 23.78 75.81 L 24.01 75.99 L 24.24 76.17 L 24.47 76.35 L 24.64 76.49 Z",
    "M 16.31 82.6 L 16.09 82.41 L 15.8 82.15 L 15.57 81.94 L 15.44 81.83 L 15.05 81.47 L 14.84 81.27 L 14.71 81.15 L 14.41 80.87 L 14.12 80.59 L 21.2 73.51 L 21.31 73.63 L 21.43 73.75 L 21.74 74.03 L 21.95 74.23 L 21.98 74.27 L 22.29 74.54 L 22.5 74.73 L 22.63 74.85 L 22.85 75.04 Z",
    "M 14.12 80.59 L 13.91 80.38 L 13.79 80.26 L 13.42 79.88 L 13.29 79.76 L 13.01 79.46 L 12.81 79.25 L 12.61 79.04 L 12.33 78.73 L 12.14 78.52 L 19.55 71.81 L 19.75 72.03 L 19.85 72.14 L 20.05 72.35 L 20.25 72.56 L 20.37 72.68 L 20.66 72.98 L 20.69 73.02 L 20.99 73.3 L 21.2 73.51 Z",
    "M 12.14 78.52 L 11.94 78.3 L 11.75 78.09 L 11.48 77.78 L 11.29 77.56 L 11.1 77.34 L 10.76 76.92 L 10.73 76.89 L 10.39 76.47 L 10.25 76.3 L 18.05 70.05 L 18.27 70.31 L 18.29 70.35 L 18.64 70.76 L 18.66 70.8 L 18.85 71.02 L 19.04 71.24 L 19.15 71.37 L 19.36 71.6 L 19.55 71.81 Z",
    "M 10.25 76.3 L 10.03 76.02 L 9.85 75.79 L 9.67 75.56 L 9.49 75.33 L 9.24 74.99 L 9.01 74.67 L 8.9 74.52 L 8.67 74.2 L 8.5 73.96 L 16.66 68.18 L 16.83 68.42 L 16.94 68.58 L 17.17 68.89 L 17.28 69.05 L 17.37 69.17 L 17.55 69.4 L 17.73 69.63 L 17.91 69.86 L 18.05 70.05 Z",
    "M 8.5 73.96 L 8.33 73.72 L 8.09 73.36 L 7.93 73.12 L 7.84 73 L 7.6 72.64 L 7.32 72.19 L 7.29 72.15 L 7.01 71.7 L 6.89 71.52 L 15.39 66.25 L 15.59 66.56 L 15.61 66.61 L 15.9 67.05 L 15.92 67.1 L 16 67.22 L 16.25 67.58 L 16.41 67.82 L 16.49 67.94 L 16.66 68.18 Z",
    "M 6.89 71.52 L 6.7 71.2 L 6.55 70.95 L 6.34 70.57 L 6.26 70.45 L 6.05 70.07 L 5.97 69.95 L 5.72 69.49 L 5.55 69.18 L 5.45 68.98 L 14.25 64.24 L 14.43 64.56 L 14.52 64.75 L 14.55 64.81 L 14.77 65.19 L 14.84 65.31 L 15.06 65.69 L 15.13 65.81 L 15.28 66.06 L 15.39 66.25 Z",
    "M 5.45 68.98 L 5.31 68.72 L 5.11 68.33 L 4.98 68.07 L 4.85 67.81 L 4.72 67.55 L 4.53 67.15 L 4.47 67.03 L 4.25 66.55 L 4.14 66.33 L 13.25 62.19 L 13.39 62.49 L 13.41 62.55 L 13.61 62.95 L 13.66 63.07 L 13.79 63.33 L 13.92 63.59 L 14.05 63.85 L 14.11 63.98 L 14.25 64.24 Z",
    "M 4.14 66.33 L 4.01 66.02 L 3.92 65.83 L 3.72 65.35 L 3.66 65.22 L 3.49 64.81 L 3.38 64.54 L 3.22 64.12 L 3.17 64 L 3.01 63.58 L 12.39 60.1 L 12.43 60.22 L 12.6 60.64 L 12.64 60.76 L 12.75 61.03 L 12.8 61.16 L 12.98 61.57 L 13 61.63 L 13.15 61.96 L 13.25 62.19 Z",
    "M 3.01 63.58 L 2.91 63.31 L 2.79 62.97 L 2.71 62.76 L 2.61 62.49 L 2.45 62 L 2.4 61.87 L 2.28 61.5 L 2.13 61.01 L 2.1 60.91 L 11.6 57.8 L 11.75 58.27 L 11.76 58.34 L 11.82 58.51 L 11.97 58.94 L 11.99 59.01 L 12.09 59.28 L 12.21 59.61 L 12.29 59.83 L 12.39 60.1 Z",
    "M 2.1 60.91 L 2.02 60.67 L 1.87 60.18 L 1.79 59.9 L 1.71 59.62 L 1.6 59.18 L 1.56 59.06 L 1.44 58.57 L 1.38 58.33 L 1.31 58.05 L 11.01 55.63 L 11.08 55.91 L 11.16 56.23 L 11.18 56.32 L 11.3 56.76 L 11.33 56.88 L 11.41 57.16 L 11.49 57.44 L 11.5 57.51 L 11.6 57.8 Z",
    "M 1.31 58.05 L 1.24 57.77 L 1.13 57.28 L 1.11 57.2 L 1 56.71 L 0.95 56.47 L 0.86 55.98 L 0.84 55.9 L 0.75 55.41 L 0.71 55.14 L 10.55 53.41 L 10.61 53.71 L 10.62 53.8 L 10.72 54.28 L 10.73 54.37 L 10.8 54.69 L 10.81 54.78 L 10.93 55.26 L 10.94 55.35 L 11.01 55.63 Z",
    "M 0.71 55.14 L 0.65 54.84 L 0.6 54.55 L 0.54 54.13 L 0.51 53.98 L 0.45 53.52 L 0.41 53.23 L 0.36 52.77 L 0.34 52.65 L 0.29 52.21 L 10.23 51.17 L 10.24 51.29 L 10.3 51.75 L 10.31 51.87 L 10.35 52.16 L 10.37 52.28 L 10.44 52.71 L 10.46 52.85 L 10.51 53.14 L 10.55 53.41 Z",
    "M 0.29 52.21 L 0.26 51.91 L 0.23 51.62 L 0.2 51.33 L 0.16 50.87 L 0.15 50.75 L 0.11 50.29 L 0.09 50 L 0.07 49.54 L 0.05 49.34 L 10.04 48.83 L 10.07 49.2 L 10.07 49.32 L 10.09 49.61 L 10.09 49.73 L 10.14 50.19 L 10.14 50.31 L 10.17 50.6 L 10.2 50.89 L 10.23 51.17 Z",
    "M 0.05 49.34 L 0.04 48.96 L 0.03 48.67 L 0.02 48.37 L 0.01 48.08 L 0 47.79 L 0 46.38 L 10 46.55 L 10 47.45 L 10.01 47.74 L 10.02 48.03 L 10.03 48.33 L 10.04 48.62 L 10.04 48.83 Z",
    "M 0 46.38 L 0.01 46 L 0.01 45.88 L 0.02 45.42 L 0.03 45.13 L 0.06 44.67 L 0.06 44.55 L 0.09 44.09 L 0.11 43.8 L 0.13 43.5 L 10.11 44.18 L 10.09 44.46 L 10.07 44.77 L 10.06 44.89 L 10.04 45.35 L 10.03 45.47 L 10.02 45.76 L 10.01 45.88 L 10.01 46.34 L 10 46.55 Z",
    "M 0.13 43.5 L 0.15 43.21 L 0.17 42.92 L 0.22 42.45 L 0.25 42.18 L 0.28 41.89 L 0.31 41.6 L 0.37 41.14 L 0.38 41.02 L 0.44 40.56 L 10.34 41.92 L 10.32 42.04 L 10.27 42.5 L 10.25 42.62 L 10.22 42.91 L 10.19 43.2 L 10.16 43.51 L 10.15 43.6 L 10.13 43.89 L 10.11 44.18 Z",
    "M 0.44 40.56 L 0.48 40.27 L 0.52 39.96 L 0.59 39.53 L 0.64 39.24 L 0.66 39.12 L 0.74 38.64 L 0.82 38.22 L 0.85 38.07 L 0.94 37.63 L 10.73 39.69 L 10.69 39.83 L 10.62 40.24 L 10.58 40.4 L 10.56 40.48 L 10.5 40.94 L 10.45 41.23 L 10.42 41.38 L 10.38 41.63 L 10.34 41.92 Z",
    "M 0.94 37.63 L 1 37.33 L 1.05 37.08 L 1.12 36.76 L 1.17 36.51 L 1.29 36.03 L 1.36 35.75 L 1.43 35.47 L 1.49 35.22 L 1.59 34.82 L 11.25 37.4 L 11.21 37.56 L 11.13 37.89 L 11.06 38.17 L 10.99 38.45 L 10.97 38.53 L 10.9 38.86 L 10.85 39.1 L 10.78 39.43 L 10.73 39.69 Z",
    "M 1.59 34.82 L 1.69 34.46 L 1.77 34.18 L 1.85 33.9 L 1.93 33.62 L 2.07 33.18 L 2.12 33.02 L 2.24 32.63 L 2.33 32.35 L 2.43 32.05 L 11.93 35.16 L 11.85 35.41 L 11.76 35.69 L 11.7 35.86 L 11.59 36.24 L 11.55 36.36 L 11.47 36.64 L 11.39 36.92 L 11.31 37.2 L 11.25 37.4 Z",
    "M 2.43 32.05 L 2.51 31.8 L 2.68 31.32 L 2.76 31.1 L 2.88 30.77 L 2.98 30.5 L 3.08 30.23 L 3.23 29.86 L 3.29 29.68 L 3.46 29.26 L 12.72 33.04 L 12.67 33.16 L 12.53 33.52 L 12.46 33.71 L 12.36 33.98 L 12.26 34.25 L 12.18 34.46 L 12.06 34.8 L 12.03 34.86 L 11.93 35.16 Z",
    "M 3.46 29.26 L 3.57 28.99 L 3.78 28.52 L 3.8 28.46 L 3.98 28.05 L 4.06 27.87 L 4.21 27.52 L 4.44 27.05 L 4.46 26.99 L 4.66 26.59 L 13.66 30.93 L 13.6 31.05 L 13.38 31.53 L 13.35 31.58 L 13.26 31.77 L 13.12 32.11 L 13.06 32.24 L 12.86 32.72 L 12.83 32.77 L 12.72 33.04 Z",
    "M 4.66 26.59 L 4.74 26.4 L 4.94 26 L 5.07 25.74 L 5.28 25.35 L 5.34 25.22 L 5.59 24.77 L 5.69 24.58 L 5.78 24.39 L 6.02 23.97 L 14.71 28.91 L 14.66 29.01 L 14.49 29.32 L 14.31 29.65 L 14.28 29.7 L 14.08 30.09 L 14.01 30.22 L 13.88 30.48 L 13.82 30.6 L 13.66 30.93 Z",
    "M 6.02 23.97 L 6.15 23.75 L 6.37 23.37 L 6.44 23.25 L 6.66 22.87 L 6.81 22.62 L 6.96 22.37 L 7.2 21.99 L 7.32 21.8 L 7.54 21.46 L 15.91 26.93 L 15.8 27.1 L 15.62 27.39 L 15.54 27.51 L 15.39 27.76 L 15.24 28.01 L 15.16 28.13 L 14.95 28.51 L 14.87 28.63 L 14.71 28.91 Z",
    "M 7.54 21.46 L 7.72 21.19 L 7.88 20.95 L 7.99 20.77 L 8.29 20.34 L 8.43 20.14 L 8.62 19.87 L 8.79 19.63 L 8.96 19.39 L 9.19 19.08 L 17.23 25.02 L 17.12 25.17 L 16.95 25.41 L 16.78 25.65 L 16.63 25.86 L 16.45 26.12 L 16.41 26.17 L 16.2 26.49 L 16.04 26.73 L 15.91 26.93 Z",
    "M 9.19 19.08 L 9.36 18.85 L 9.62 18.51 L 9.74 18.35 L 9.89 18.15 L 10.32 17.63 L 10.34 17.59 L 10.52 17.36 L 10.77 17.05 L 11.02 16.77 L 18.64 23.23 L 18.51 23.39 L 18.4 23.52 L 18.22 23.75 L 17.88 24.17 L 17.93 24.09 L 17.74 24.35 L 17.5 24.67 L 17.4 24.79 L 17.23 25.02 Z",
    "M 11.02 16.77 L 11.17 16.59 L 11.43 16.28 L 11.62 16.06 L 11.88 15.77 L 12.09 15.54 L 12.2 15.41 L 12.56 15.02 L 12.76 14.81 L 12.92 14.63 L 20.24 21.45 L 20 21.71 L 19.8 21.92 L 19.76 21.95 L 19.49 22.26 L 19.3 22.47 L 19.18 22.6 L 18.99 22.82 L 18.87 22.95 L 18.64 23.23 Z",
    "M 12.92 14.63 L 13.16 14.38 L 13.44 14.09 L 13.56 13.97 L 13.85 13.67 L 14.15 13.39 L 14.36 13.19 L 14.48 13.06 L 14.78 12.78 L 15.04 12.54 L 21.84 19.88 L 21.68 20.02 L 21.56 20.14 L 21.26 20.43 L 21.05 20.63 L 20.93 20.75 L 20.8 20.87 L 20.52 21.17 L 20.4 21.28 L 20.24 21.45 Z",
    "M 15.04 12.54 L 15.2 12.39 L 15.51 12.11 L 15.73 11.91 L 16.03 11.65 L 16.25 11.46 L 16.47 11.27 L 16.69 11.08 L 17.01 10.82 L 17.23 10.63 L 23.59 18.35 L 23.37 18.52 L 23.23 18.64 L 23.01 18.83 L 22.79 19.02 L 22.57 19.21 L 22.43 19.33 L 22.23 19.51 L 22.1 19.63 L 21.84 19.88 Z",
    "M 17.23 10.63 L 17.46 10.45 L 17.69 10.25 L 18.01 10 L 18.24 9.82 L 18.47 9.64 L 18.81 9.39 L 19.02 9.23 L 19.28 9.04 L 19.56 8.84 L 25.42 16.94 L 25.22 17.08 L 25.02 17.23 L 24.75 17.43 L 24.63 17.52 L 24.4 17.7 L 24.17 17.88 L 24.03 17.99 L 23.82 18.15 L 23.59 18.35 Z",
    "M 19.56 8.84 L 19.75 8.7 L 20.19 8.4 L 20.31 8.31 L 20.67 8.07 L 20.79 7.98 L 21.15 7.74 L 21.47 7.53 L 21.76 7.35 L 21.98 7.21 L 27.34 15.66 L 27.06 15.83 L 26.87 15.95 L 26.69 16.06 L 26.57 16.14 L 26.21 16.39 L 26.09 16.47 L 25.73 16.72 L 25.69 16.74 L 25.42 16.94 Z",
    "M 21.98 7.21 L 22.25 7.04 L 22.45 6.91 L 22.83 6.68 L 23.08 6.53 L 23.46 6.32 L 23.58 6.24 L 24.04 5.99 L 24.09 5.95 L 24.58 5.68 L 29.26 14.52 L 29.23 14.53 L 28.78 14.79 L 28.72 14.82 L 28.34 15.04 L 28.22 15.11 L 27.97 15.26 L 27.85 15.33 L 27.55 15.52 L 27.34 15.66 Z",
    "M 24.58 5.68 L 24.86 5.53 L 25.06 5.43 L 25.32 5.29 L 25.63 5.12 L 25.96 4.96 L 26.22 4.83 L 26.56 4.66 L 26.89 4.51 L 27.08 4.42 L 31.42 13.43 L 31.09 13.59 L 30.9 13.68 L 30.7 13.77 L 30.44 13.9 L 30.25 14 L 30.06 14.09 L 29.8 14.23 L 29.48 14.41 L 29.26 14.52 Z",
    "M 27.08 4.42 L 27.41 4.26 L 27.75 4.11 L 28.02 3.99 L 28.36 3.85 L 28.55 3.76 L 28.96 3.59 L 29.23 3.48 L 29.5 3.37 L 29.84 3.23 L 33.47 12.55 L 33.28 12.63 L 33.01 12.74 L 32.74 12.85 L 32.61 12.9 L 32.26 13.05 L 32.08 13.13 L 31.81 13.25 L 31.61 13.34 L 31.42 13.43 Z",
    "M 29.84 3.23 L 30.19 3.1 L 30.31 3.05 L 30.73 2.89 L 31 2.79 L 31.34 2.67 L 31.71 2.55 L 31.89 2.48 L 32.26 2.36 L 32.59 2.25 L 35.65 11.77 L 35.42 11.84 L 35.25 11.9 L 34.87 12.03 L 34.7 12.09 L 34.48 12.17 L 34.21 12.27 L 34.09 12.31 L 33.67 12.48 L 33.47 12.55 Z",
    "M 32.59 2.25 L 32.87 2.16 L 33.31 2.03 L 33.37 2.01 L 33.86 1.86 L 34.14 1.78 L 34.42 1.7 L 34.86 1.59 L 34.98 1.55 L 35.45 1.44 L 37.83 11.14 L 37.72 11.17 L 37.28 11.29 L 37.16 11.32 L 36.88 11.4 L 36.6 11.48 L 36.53 11.49 L 36.05 11.65 L 35.93 11.68 L 35.65 11.77 Z",
    "M 35.45 1.44 L 35.71 1.37 L 35.99 1.3 L 36.27 1.23 L 36.76 1.12 L 37 1.07 L 37.17 1.03 L 37.74 0.92 L 37.9 0.88 L 38.24 0.81 L 40.14 10.63 L 39.92 10.68 L 39.5 10.76 L 39.51 10.75 L 39.1 10.85 L 38.78 10.92 L 38.69 10.93 L 38.41 11 L 38.13 11.07 L 37.83 11.14 Z",
    "M 38.24 0.81 L 38.63 0.74 L 38.88 0.7 L 39.2 0.64 L 39.66 0.58 L 39.78 0.55 L 40.2 0.49 L 40.52 0.45 L 40.81 0.41 L 41.19 0.36 L 42.38 10.28 L 42.17 10.31 L 41.88 10.35 L 41.62 10.39 L 41.48 10.41 L 41.02 10.48 L 40.9 10.5 L 40.64 10.54 L 40.33 10.6 L 40.14 10.63 Z",
    "M 41.19 0.36 L 41.56 0.32 L 41.68 0.3 L 42.11 0.25 L 42.42 0.22 L 42.88 0.18 L 43 0.17 L 43.46 0.13 L 43.75 0.11 L 44.04 0.09 L 44.72 10.07 L 44.43 10.09 L 44.14 10.11 L 44.02 10.11 L 43.56 10.16 L 43.44 10.16 L 43.17 10.19 L 43.04 10.2 L 42.58 10.26 L 42.38 10.28 Z",
    // "M 44.04 0.09 L 44.33 0.07 L 44.79 0.05 L 45.09 0.04 L 45.21 0.03 L 45.84 0.02 L 45.96 0.01 L 46.25 0 L 46.71 0 L 48.25 0.03 L 47.91 10.02 L 46.71 10 L 46.59 10 L 46.3 10.01 L 45.84 10.02 L 45.89 10.01 L 45.43 10.04 L 45.13 10.05 L 45.01 10.05 L 44.72 10.07 Z",
    // "M 48 0 C 50.761 0 53 2.239 53 5 C 53 7.761 50.761 10 48 10 C 45.239 10 43 7.761 43 5 C 43 2.239 45.239 0 48 0 Z",
    "M 44.04 0.09 L 44.33 0.07 L 44.79 0.05 L 45.09 0.04 L 45.21 0.03 L 45.84 0.02 L 45.96 0.01 L 46.25 0 L 46.71 0 L 47.606 0.017 C 47.727 0.006 47.858 0 48 0 C 50.761 0 53 2.239 53 5 C 53 7.761 50.761 10 48 10 C 47.97 10 47.94 10 47.911 9.999 L 47.91 10.02 L 46.71 10 L 46.59 10 L 46.3 10.01 L 45.84 10.02 L 45.89 10.01 L 45.43 10.04 L 45.13 10.05 L 45.01 10.05 L 44.72 10.07 Z",
]
